### PR TITLE
feat: uid-mux

### PR DESCRIPTION
**Pull request by sinui0**
This PR moves/reimplements `uid-mux` in this repo.

# Changes
- Uses newer version of `yamux` which has better flow control. This should give us higher throughput.
- Stream ids are now hashed to a fixed-length
- Better logging
- Cleaner.

### CODE REVIEW

**Comment by sinui0** [comment]
This is supporting work to implement a multi-threaded executor in `mpz`

**Comment by sinui0** [comment]
Sigh, hold off on review. There is a deadlock somewhere

**Comment by sinui0** [comment]
It is seeming to me that `yamux` may have a bug, hard to say conclusively. It looks like their implementation is letting data sit in the write buffer with no way to flush it.

**Comment by sinui0** [comment]
Nope, nevermind there is a way to flush it. Fix coming soon

**Comment by sinui0** [comment]
@th4s @themighty1 ready!

File: uid-mux/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        Ok(())
+    }
+}
+
+impl AsRef<[u8]> for InternalId {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
```
Comment (th4s) [inline uid-mux/src/lib.rs:48]: I would suggest to replace/add
```rust
impl Borrow<[u8; 32]> for InternalId {
    fn borrow(&self) -> &[u8; 32] {
        &self.0
    }
}
```

File: uid-mux/src/yamux.rs [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    }
+}
+
+/// A yamux multiplexer.
+#[derive(Debug)]
+pub struct Yamux<Io> {
+    role: Role,
+    conn: Connection<Io>,
+    queue: Arc<Mutex<Queue>>,
+    close_notify: Arc<Notify>,
```
Comment (th4s) [inline uid-mux/src/yamux.rs:53]: Any chance to do this executor agnostic and not use `tokio::sync::Notify`?

File: uid-mux/src/yamux.rs [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    shutdown_notify: Arc<AtomicBool>,
```
Comment (th4s) [inline uid-mux/src/yamux.rs:54]: Is the `Arc` for the atomic to work around lifetimes?

File: uid-mux/src/yamux.rs [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+        Ok(())
+    }
+
+    fn server_process_inbound(&mut self, cx: &mut Context<'_>) -> Result<()> {
+        while let Poll::Ready(Some((id, stream))) =
+            self.incoming.poll_next_unpin(cx).map(Option::transpose)?
+        {
+            debug!("received stream: {}", id);
+            let mut queue = self.queue.lock().unwrap();
```
Comment (themighty1) [inline uid-mux/src/yamux.rs]: Shouldn't this be locked outside of the while loop?

File: uid-mux/src/yamux.rs [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+        if let Poll::Ready(stream) = self.conn.poll_next_inbound(cx).map(Option::transpose)? {
+            if stream.is_some() {
+                error!("client mux received incoming stream");
+                return Err(
+                    std::io::Error::other("client mode can not accept incoming streams").into(),
+                );
+            }
+
+            info!("remote closed connection");
+            self.remote_closed = true;
```
Comment (th4s) [inline uid-mux/src/yamux.rs:164]: Why do we know that remote closed the connection here?

File: uid-mux/src/yamux.rs [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+    async fn test_yamux() {
+        let (client_io, server_io) = duplex(1024);
+        let client = Yamux::new(client_io.compat(), Config::default(), Mode::Client);
+        let server = Yamux::new(server_io.compat(), Config::default(), Mode::Server);
+
+        let client_ctrl = client.control();
+        let server_ctrl = server.control();
+
+        let conn_task = tokio::spawn(async {
+            futures::try_join!(client.into_future(), server.into_future()).unwrap();
```
Comment (themighty1) [inline uid-mux/src/yamux.rs:422]: Would it be possible to test 2 situations disctinctly:
- the future resolves because of ctrl.close()
- the future resolves because the remote called its ctrl.close()

right now it's not obvious why it resolves

File: uid-mux/src/yamux.rs [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+        }
+
+        while let Poll::Ready(Some(())) =
+            self.outgoing.poll_next_unpin(cx).map(Option::transpose)?
+        {
+            trace!("finished processing stream");
+        }
+
+        // Set the waker to wake up when the queue is ready to process more streams.
+        queue.waker = cx.waker().clone();
```
Comment (themighty1) [inline uid-mux/src/yamux.rs]: is there a reson why this is needed here? we will set the waker below with  self.set_queue_waker(cx);

Comment (th4s) [inline uid-mux/src/yamux.rs]: Why is this needed? Doing it in `YamuxFuture::poll` should be sufficient, like you already do.

File: uid-mux/src/yamux.rs [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+    queue: Arc<Mutex<Queue>>,
+    close_notify: Arc<Notify>,
+    shutdown_notify: Arc<AtomicBool>,
+}
+
+#[derive(Debug)]
+struct Queue {
+    waiting: HashMap<InternalId, oneshot::Sender<Stream>>,
+    ready: HashMap<InternalId, Stream>,
+    waker: Waker,
```
Comment (themighty1) [inline uid-mux/src/yamux.rs]: In my mind the waker causes the future to be polled, so it's more fitting to be a field in `YamuxFuture`.
I have a hard time mapping waker to the `Queue`.
In the code we wake the Future in order to close the connection, regardless if there is anything in the Queue, so i don't see a direct connection between waker and `Queue`.

**Review by themighty1 — CHANGES_REQUESTED** [review]
gw!!
Some comments, potentially blocking.

Also, could you pls adjust  the comments acc.to the style guide to be /// Create --> Creates etc.
I'll open a PR to add the guide to this repo.

**Review by th4s — COMMENTED** [review]
Very nice and clean :rocket:

There is one thing I do not get. There has to be some future running in order to transmit the data between client and server on existing streams. But currently it looks to me that in `YamuxFuture::poll` in the end we return `Poll::Pending`. So there has to be some way to wake it again, so that again data is transmitted.

But the only place we wake it, is when opening a stream (and when closing the connection). This to me sounds like data on existing streams is only transmitted when a new stream is opened or the connection is closed. 

So what I mean is that `yamux::Connection::poll_next_inbound` is not permanently polled.

EDIT: After some more investigation, I now suspect that it works correctly, because `yamux::Connection::Active` contains a `loop` in its `poll` method.

File: uid-mux/src/yamux.rs (hunk 3 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs:54]: Yes, `AtomicBool` is not `Clone`

File: uid-mux/src/yamux.rs (hunk 2 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs:53]: `Notify` is not executor specific

File: uid-mux/src/lib.rs (hunk 1 above)
Comment (sinui0) [reply uid-mux/src/lib.rs:48]: For what purpose?

File: uid-mux/src/yamux.rs (hunk 5 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs:164]: When it returns `Poll::Ready(None)` that indicates the connection has closed

File: uid-mux/src/yamux.rs (hunk 8 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs]: `Waker` is a trigger which will cause the executor to poll `YamuxFuture`

> i don't see a direct connection between waker and Queue.

When `YamuxCtrl` updates the queue it needs to be able to trigger `YamuxFuture` to be polled so that it will process anything in `waiting`. It doesn't make much sense for a future to own the thing that is needed to wake itself up.

**Comment by sinui0** [comment]
@th4s 

> But the only place we wake it, is when opening a stream (and when closing the connection)

Anytime a `Future` returns `Poll::Pending` it (should) clone the `Waker` provided by `Context` and ensure that it is triggered when it is ready to be polled again. All the polls of the underlying yamux connection are doing this, so they contribute to driving `YamuxFuture` in addition to `YamuxCtrl`

**Comment by sinui0** [comment]
@th4s @themighty1 it seems my understanding of `Waker` was not complete. ~~Apparently it is re-usable across multiple `Future:poll` calls and does not need to be cloned multiple times. This is nice and simplifies things a bit.~~

Edit: Nevermind, ChatGPT lied to me! :joy: 

[Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup.](https://doc.rust-lang.org/stable/std/future/trait.Future.html#the-poll-method)

File: uid-mux/src/yamux.rs (hunk 6 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs:422]: Updated to be more comprehensive.

File: uid-mux/src/yamux.rs (hunk 4 above)
Comment (sinui0) [reply uid-mux/src/yamux.rs]: Yeah

**Review by th4s — APPROVED** [review]

File: uid-mux/src/yamux.rs (hunk 2 above)
Comment (th4s) [reply uid-mux/src/yamux.rs:53]: Ah right, but you have to pull in `tokio`. But you only pull in `sync` feature, so should be ok.

File: uid-mux/src/lib.rs (hunk 1 above)
Comment (th4s) [reply uid-mux/src/lib.rs:48]: `Borrow` is stronger than `AsRef`, but I now see that it wouldn't offer any specific advantages here. It's ok let's keep it.

File: uid-mux/src/yamux.rs [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+where
+    Io: AsyncWrite + AsyncRead + Unpin,
+{
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip_all, err))]
+    fn client_handle_inbound(&mut self, cx: &mut Context<'_>) -> Result<()> {
+        if let Poll::Ready(stream) = self.conn.poll_next_inbound(cx).map(Option::transpose)? {
+            if stream.is_some() {
+                error!("client mux received incoming stream");
+                return Err(
+                    std::io::Error::other("client mode can not accept incoming streams").into(),
```
Comment (themighty1) [inline uid-mux/src/yamux.rs]:
```suggestion
                    std::io::Error::other("client mode cannot accept incoming streams").into(),
```

File: uid-mux/src/yamux.rs [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+#[derive(Debug, Clone)]
+pub struct YamuxCtrl {
+    role: Role,
+    queue: Arc<Mutex<Queue>>,
+    close_notify: Arc<Notify>,
+    shutdown_notify: Arc<AtomicBool>,
+}
+
+impl YamuxCtrl {
+    /// Close the yamux connection.
```
Comment (themighty1) [inline uid-mux/src/yamux.rs]:
```suggestion
    /// Closes the yamux connection.
```

**Review by themighty1 — CHANGES_REQUESTED** [review]

**Comment by themighty1** [comment]
There is a race condition iiuc which violates this requirement: "Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup"

Suppose that the future has just been polled and the execution is somewhere around this line https://github.com/tlsnotary/tlsn-utils/blob/e2b207c2528d2bae167af698112c9c07777e09db/uid-mux/src/yamux.rs#L309-L310

If around the same time the `YamuxCtrl::close()` call happens  and tries to get the waker in https://github.com/tlsnotary/tlsn-utils/blob/e2b207c2528d2bae167af698112c9c07777e09db/uid-mux/src/yamux.rs#L345
it will get the waker from the **previous** context, not from the **current** context as per the requirement.

**Comment by sinui0** [comment]
That would be fine, as the future would already be getting polled. What matters is that the waker is reset before the future returns `Poll::Pending`, which it does on every poll. The risk of waking a stale waker is that the future never gets polled again and you get a deadlock.

**Review by themighty1 — APPROVED** [review]
ack
